### PR TITLE
 Add Library Guide for User Defined Functions: Window/Aggregate

### LINKED ISSUE #7307: Library Guide: Adding User Defined Functions: Scalar/Window/Aggregate/
### Is your feature request related to a problem or challenge?

Part of https://github.com/apache/arrow-datafusion/issues/7014

If we want to have DataFusion used as the core of many new systems, we need it to be as easy as possible for someone to get their idea working on top of DataFusion.

Thanks to @tshauck we now have a basic Library Users Guide ❤️ and this ticket describes expanding it out

### Describe the solution you'd like

Fill in the content of https://arrow.apache.org/datafusion/library-user-guide/adding-udfs.html

Example Outline
1. How to create and register a `ScalarUDF`
1. How to create and register a `AggregateUDF`
1. How to create and register a `WindowUDF`

The examples directory holds a bunch more of examples: https://github.com/apache/arrow-datafusion/tree/main/datafusion-examples

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Veeupup**
## Which issue does this PR close?

Closes #7307 .

## What changes are included in this PR?

Add doc for user-defined window/aggregate function

### CODE REVIEW

**Comment by andygrove** [comment]
Thanks for contributing more documentation @Veeupup

File: docs/source/library-user-guide/adding-udfs.md [hunk 1]
```diff
@@ -115,10 +122,313 @@ let df = ctx.sql(&sql).await.unwrap();
...
+                self.prod *= value;
+                self.n += 1;
+            } else {
+                unreachable!("")
+            }
+            Ok(())
+        })
+    }
+
+    // Optimization hint: this trait also supports `update_batch` and `merge_batch`,
```
Comment (alamb) [inline docs/source/library-user-guide/adding-udfs.md:357]: this comment seems out of date as it is on `merge_batch`

**Review by alamb — APPROVED** [review]
Thank you very much @Veeupup  -- this is super helpful and a really nice PR 🙏  thank you for this (and all your other recent, high quality contributions ❤️ )

Something related to this that you might be interested in is https://github.com/apache/arrow-datafusion/issues/6747 (and I had a POC version of ithttps://github.com/apache/arrow-datafusion/pull/6746) to help make window functions more easily and ergonomically (creating `Expr`s programatically rather than via SQL).

**Comment by alamb** [comment]
While I am sure there are improvements we can make to this documentation, this is a lot better than we have now, so let's merge it in and iterate in subequent PRs. Thanks again @Veeupup

**Pull request merged**